engines/freescape/gfx_tinygl.cpp

```diff
@@ -0,0 +1,586 @@
+
+	tglMatrixMode(TGL_PROJECTION);
+	tglLoadIdentity();
+
+	tglMatrixMode(TGL_MODELVIEW);
+	tglLoadIdentity();
+
+	tglDisable(TGL_LIGHTING);
+	tglDisable(TGL_TEXTURE_2D);
+	tglEnable(TGL_DEPTH_TEST);
+}
+
+void TinyGLRenderer::setViewport(const Common::Rect &rect) {
+	tglViewport(rect.left, g_system->getHeight() - rect.bottom, rect.width(), rect.height());
+}
+
+void TinyGLRenderer::clear() {
+	tglClear(TGL_COLOR_BUFFER_BIT | TGL_DEPTH_BUFFER_BIT);
+	tglColor3f(1.0f, 1.0f, 1.0f);
+}
+
+void TinyGLRenderer::drawTexturedRect2D(const Common::Rect &screenRect, const Common::Rect &textureRect, Texture *texture) {
+	const float sLeft = screenRect.left;
+	const float sTop = screenRect.top;
+	const float sWidth = screenRect.width();
+	const float sHeight = screenRect.height();
+
+	// HACK: tglBlit is not affected by the viewport, so we offset the draw coordinates here
+	int viewPort[4];
+	tglGetIntegerv(TGL_VIEWPORT, viewPort);
+
+	TinyGL::BlitTransform transform(sLeft + viewPort[0], sTop + viewPort[1]);
+	transform.sourceRectangle(textureRect.left, textureRect.top, sWidth, sHeight);
+	tglBlit(((TinyGLTexture *)texture)->getBlitTexture(), transform);
+}
+
+void TinyGLRenderer::updateProjectionMatrix(float fov, float nearClipPlane, float farClipPlane) {
+	tglMatrixMode(TGL_PROJECTION);
+	tglLoadIdentity();
+
+	float aspectRatio = _screenW / (float)_screenH;
+
+	float xmaxValue = nearClipPlane * tan(Common::deg2rad(fov) / 2);
+	float ymaxValue = xmaxValue / aspectRatio;
+	// debug("max values: %f %f", xmaxValue, ymaxValue);
+
+	tglFrustum(xmaxValue, -xmaxValue, -ymaxValue, ymaxValue, nearClipPlane, farClipPlane);
+	tglMatrixMode(TGL_MODELVIEW);
+	tglLoadIdentity();
+}
+
+void TinyGLRenderer::positionCamera(const Math::Vector3d &pos, const Math::Vector3d &interest) {
+	Math::Vector3d up_vec(0, 1, 0);
+
+	Math::Matrix4 lookMatrix = Math::makeLookAtMatrix(pos, interest, up_vec);
+	tglMultMatrixf(lookMatrix.getData());
+	tglTranslatef(-pos.x(), -pos.y(), -pos.z());
+}
+
+void TinyGLRenderer::drawRect2D(const Common::Rect &rect, uint8 a, uint8 r, uint8 g, uint8 b) {
+	tglDisable(TGL_TEXTURE_2D);
+	tglColor4ub(r, g, b, a);
+
+	if (a != 255) {
+		tglEnable(TGL_BLEND);
+		tglBlendFunc(TGL_SRC_ALPHA, TGL_ONE_MINUS_SRC_ALPHA);
+	}
+
+	tglBegin(TGL_TRIANGLE_STRIP);
+	tglVertex3f(rect.left, rect.bottom, 0.0f);
+	tglVertex3f(rect.right, rect.bottom, 0.0f);
+	tglVertex3f(rect.left, rect.top, 0.0f);
+	tglVertex3f(rect.right, rect.top, 0.0f);
+	tglEnd();
+
+	tglDisable(TGL_BLEND);
+}
+
+void TinyGLRenderer::renderCrossair(byte color, const Common::Point position) {
+	uint8 r, g, b;
+	_palette->getRGBAt(color, r, g, b); // TODO: should use opposite color
+
+	tglMatrixMode(TGL_PROJECTION);
+	tglLoadIdentity();
+	tglOrtho(0, _screenW, _screenH, 0, 0, 1);
+	tglMatrixMode(TGL_MODELVIEW);
+	tglLoadIdentity();
+
+	tglDisable(TGL_DEPTH_TEST);
+	tglDepthMask(TGL_FALSE);
+
+	tglColor3ub(r, g, b);
+
+	tglBegin(TGL_LINES);
```

Comment:
I suggest to replace tglBegin/tglEnd with tglDrawArrays/tglDrawElements.
My idea for renderers is to provide GLES1 support. So preferable is engines do not use it any more. For TinyGL is not important, but I think it will align between TinyGL and classic GL renderer.

Reply:
If I understand correctly, I should use: `glDrawArrays(TGL_LINES, 0, 2)` here?

Reply:
yes

Reply:
This should work, but wanted to check with you before changing all the API:

```c++
static TGLfloat verts[20][3] = { 
	//X   Y     Z
	0.0,  0.0,  0.0,
	0.0,  0.0,  0.0,
	...
};

void TinyGLRenderer::renderFace(const Common::Array<Math::Vector3d> &vertices) {
	assert(vertices.size() >= 2);
	const Math::Vector3d &v0 = vertices[0];

	if (vertices.size() == 2) {
		const Math::Vector3d &v1 = vertices[1];
		if (v0 == v1)
			return;

		tglEnableClientState(TGL_VERTEX_ARRAY);
		verts[0][0] = v0.x(); verts[0][1] = v0.y(); verts[0][2] = v0.z();
		verts[1][0] = v1.x(); verts[1][1] = v1.y(); verts[1][2] = v1.z();
		tglVertexPointer(3, TGL_FLOAT, 0, verts);
		tglDrawArrays(TGL_LINES, 0, 2);
		tglDisableClientState(TGL_VERTEX_ARRAY);
		return;
	}

	tglEnableClientState(TGL_VERTEX_ARRAY);
	int vi = 0;
	for (uint i = 1; i < vertices.size() - 1; i++) { // no underflow since vertices.size() > 2
		const Math::Vector3d &v1 = vertices[i];
		const Math::Vector3d &v2 = vertices[i + 1];
		vi = 3 * (i - 1);
		verts[vi + 0][0] = v0.x(); verts[vi + 0][1] = v0.y(); verts[vi + 0][2] = v0.z();
		verts[vi + 1][0] = v1.x(); verts[vi + 1][1] = v1.y(); verts[vi + 1][2] = v1.z();
		verts[vi + 2][0] = v2.x(); verts[vi + 2][1] = v2.y(); verts[vi + 2][2] = v2.z();
	}
	tglVertexPointer(3, TGL_FLOAT, 0, verts);
	tglDrawArrays(TGL_TRIANGLES, 0, vi + 3);
	tglDisableClientState(TGL_VERTEX_ARRAY);
}
```

Reply:
I recommend putting verts into heap as its dynamic, allocate it at renderer creation. instead using offsets in array, I usages create some struct with x, y, z

Reply:
Great, please review https://github.com/scummvm/scummvm/pull/4370/commits/90695e42b7c24404ea1be0fdfd0091f219c33dcd so I can continue with the rest if this looks good.

Reply:
I get impression that calculations are over complicated :) but in general fine

Reply:
Your impression is right. This is caused by my poor understanding of the 3D graphics in general :disappointed:. It should improve in the near future.

Reply:
btw @aquadran, I'm considering moving all the abstract code that does not depend on TinyGL API to the renderer class (e.g. drawPyramid). Do you think it will make sense in order to add another renderer in the near future? (e.g. OpenGL)

Reply:
yes, I think it's good idea, to prevent code duplication

Reply:
tglBegin still here